romana/openstack/tenant.go

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) 2016 Pani Networks
+// All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+// Package openstack implements openstack API specific
+// helper functions.
+package openstack
+
+import (
+	"log"
+	"os"
+
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/openstack"
+	"github.com/rackspace/gophercloud/openstack/identity/v2/tenants"
+	"github.com/rackspace/gophercloud/pagination"
+)
+
+var (
+	identityClient *gophercloud.ServiceClient
+	networkClient  *gophercloud.ServiceClient
+	computeClient  *gophercloud.ServiceClient
+)
+
+type ClientInterface interface {
+}
+
+func getIdentityClient() (*gophercloud.ServiceClient, error) {
+	var err error
+	if identityClient == nil {
+		identityClient, err = initIdentityClient()
+		if err != nil {
+			log.Println("Error: ", err)
+			return nil, err
+		}
+	}
+	return identityClient, nil
+}
+
+func getComputeClient() (*gophercloud.ServiceClient, error) {
+	var err error
+	if computeClient == nil {
+		computeClient, err = initComputeClient()
+		if err != nil {
+			log.Println("Error: ", err)
+			return nil, err
+		}
+	}
+	return computeClient, nil
+}
+
+func getNetworkClient() (*gophercloud.ServiceClient, error) {
+	var err error
+	if networkClient == nil {
+		networkClient, err = initNetworkClient()
+		if err != nil {
+			log.Println("Error: ", err)
+			return nil, err
+		}
+	}
+	return networkClient, nil
+}
+
+// initIdentityClient initializes openstack api using
+// gophercloud. Currently it uses enviornmental variables
+// for authenticating with openstack identity.
+func initIdentityClient() (*gophercloud.ServiceClient, error) {
+	opts, err := openstack.AuthOptionsFromEnv()
```

Comment:
These three init... functions have most code in common. Seems like that could be re-factored.